SNOW-18856699: Add usage details (#25)
snowflakedb/floe: cpp/README.md

sudo apt-get install cmake g++ libssl-dev clang-tidy lcov
```

## Usage

### Include the Header

```cpp
#include <floe/floe.hpp>
```

### Create a Key

```cpp
#include <floe/floe.hpp>
#include <vector>
#include <span>

using namespace sf;

> // Your 32-byte key (in production, use a secure key derivation function)

Reviewer: ```suggestion
// Your 32-byte key (in production, use a randomly generated or securely derived key)
```

std::vector<ub1> rawKey(32);
// ... populate rawKey with secure random bytes ...

// Choose a parameter spec
FloeParameterSpec params = FloeParameterSpec::GCM256_IV256_4K();

// Create the key object
FloeKey key(rawKey, params);
if (!key.isValid()) {
  // Handle invalid key

> }

Reviewer: This should be omitted. It's implicitly called when you try to use the key.

Author: Removed

```

### Encryption

```cpp
// Optional additional authenticated data (AAD)
std::vector<ub1> aad = {'m', 'e', 't', 'a', 'd', 'a', 't', 'a'};
std::span<const ub1> aadSpan(aad);

// Create encryptor
auto [result, encryptor] = FloeEncryptor::create(key, aadSpan);
if (result != FloeResult::Success) {
  std::cerr << "Error: " << floeErrorMessage(result) << std::endl;
  return;
}

// Get the header (must be stored/transmitted with the ciphertext)
auto header = encryptor->getHeader();

// Build ciphertext: header || encrypted segments
std::vector<ub1> ciphertext;
ciphertext.insert(ciphertext.end(), header.begin(), header.end());

// Encrypt data in segments
std::span<const ub1> plaintext(plaintextData);
for (size_t offset = 0; offset < plaintext.size();
     offset += params.getPlaintextSegmentLength()) {

  std::vector<ub1> segment;
  bool isLastSegment = (offset + params.getPlaintextSegmentLength() >= plaintext.size());

  if (isLastSegment) {
    // Final segment (may be shorter than a full segment)
    size_t lastSegmentSize = plaintext.size() - offset;
    segment.resize(encryptor->sizeOfLastOutput(lastSegmentSize));
    std::span<ub1> outputSpan(segment);

>     result = encryptor->processLastSegment(plaintext.subspan(offset), outputSpan);

Reviewer: I believe that `std::vector` will implicitly cast to `std::span`. I'd make this change here and elsewhere.

```suggestion
    result = encryptor->processLastSegment(plaintext.subspan(offset), segment);
```

  } else {
    // Full segment
    segment.resize(params.getEncryptedSegmentLength());
    std::span<ub1> outputSpan(segment);

    result = encryptor->processSegment(

>         plaintext.subspan(offset, params.getPlaintextSegmentLength()), outputSpan);

Reviewer: ```suggestion
    result = encryptor->processSegment(
        plaintext.subspan(offset, params.getPlaintextSegmentLength()), segment);
```

  }

  if (result != FloeResult::Success) {
    std::cerr << "Encryption error: " << floeErrorMessage(result) << std::endl;
    return;
  }

  ciphertext.insert(ciphertext.end(), segment.begin(), segment.end());
}
```

### Decryption

```cpp
// ciphertext contains: header || segment1 || segment2 || ... || lastSegment
std::span<const ub1> ciphertext(ciphertextData);

// Create decryptor using the header (first part of ciphertext)
auto [result, decryptor] = FloeDecryptor::create(key, aadSpan, ciphertext);
if (result != FloeResult::Success) {
  std::cerr << "Error: " << floeErrorMessage(result) << std::endl;
  return;
}

// Decrypt segments
std::vector<ub1> decrypted;
for (size_t offset = params.getHeaderLength(); offset < ciphertext.size();
     offset += params.getEncryptedSegmentLength()) {

  std::vector<ub1> plaintext;
  bool isLastSegment = (offset + params.getEncryptedSegmentLength() >= ciphertext.size());

  if (isLastSegment) {
    // Final segment may be shorter
    size_t lastSegmentSize = ciphertext.size() - offset;
    plaintext.resize(decryptor->sizeOfLastOutput(lastSegmentSize));
    std::span<ub1> outputSpan(plaintext);

>     result = decryptor->processLastSegment(ciphertext.subspan(offset), outputSpan);

Reviewer: ```suggestion
    result = decryptor->processLastSegment(ciphertext.subspan(offset), plaintext);
```

  } else {
    // Full segment
    plaintext.resize(params.getPlaintextSegmentLength());
    std::span<ub1> outputSpan(plaintext);

    result = decryptor->processSegment(

>         ciphertext.subspan(offset, params.getEncryptedSegmentLength()), outputSpan);

Reviewer: ```suggestion
    result = decryptor->processSegment(
        ciphertext.subspan(offset, params.getEncryptedSegmentLength()), plaintext);
```

  }

  if (result != FloeResult::Success) {
    std::cerr << "Decryption error: " << floeErrorMessage(result) << std::endl;
    return;
  }

  decrypted.insert(decrypted.end(), plaintext.begin(), plaintext.end());
}

// Verify decryption completed successfully
result = decryptor->finish();
if (result != FloeResult::Success) {
  std::cerr << "Truncated ciphertext!" << std::endl;
}
```

### Parameter Specifications

| Specification | Segment Size | Use Case |
|---------------|--------------|----------|
| `GCM256_IV256_4K()` | 4 KB | General purpose, good for most files |

> | `GCM256_IV256_1M()` | 1 MB | Large files, fewer segments |

Reviewer: Honestly. I'd probably recommend the 1MB for most cases. 4KB is really only useful on memory constrained systems.

Author: Done
